### PR TITLE
Feature/typescript

### PR DESCRIPTION
**Pull request by larsmunkholm**
**Support for TypeScript**

## Does this PR introduce a breaking change?
No

I haven't changed the files under the performance-folder to TypeScipt, and I'm not sure if I should?

### CODE REVIEW

File: packages/js-config-webpack-plugin/package.json [hunk 1]
```diff
@@ -34,6 +35,7 @@
     "core-js": "3.4.1",
     "react-hot-loader": "4.12.19",
     "terser-webpack-plugin": "2.2.1",
+    "ts-loader": "^6.2.1",
```
Comment (jeppeskovsen) [inline packages/js-config-webpack-plugin/package.json]: Er ts-loader nødvendig?

Comment (larsmunkholm) [reply packages/js-config-webpack-plugin/package.json]: @jeppeskovsen Nej :) Den får jeg fjernet.. Havde også glemt at smække JS_FILE_REGEX og TS_FILE_REGEX sammen i development.config og production.config ... det er på vej!

Comment (larsmunkholm) [reply packages/js-config-webpack-plugin/package.json]: Og det er rettet nu.

**Comment by jeppeskovsen** [comment]
When this is completed, an example in the frontline-starter-spa repository would be nice :)

File: packages/js-config-webpack-plugin/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        "plugin:@typescript-eslint/recommended",
+        "plugin:prettier/recommended",
+        "prettier/@typescript-eslint"
+    ],
+    parserOptions: {
+        project: "tsconfig.json",
+        ecmaVersion: 2018,
+        sourceType: "module"
+    }
+};
```
Comment (dennishn) [inline packages/js-config-webpack-plugin/.eslintrc.js:14]: There should be no need for adding specific linting rules to our packages - they are all configured by .eslintrc.js in the root.

File: packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts [hunk 3]
```diff
@@ -13,6 +15,9 @@ export interface FrontlineJsConfigWebpackPluginOptions {
     mode: "production" | "development" | "none";
     browserslistEnv: string;
     babelConfigFile?: string;
+    tsConfigFile?: string;
+    plugins?: any;
+    enableEslint?: boolean;
```
Comment (dennishn) [inline packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts:20]: Whats the goal of this option, what happens if it is set to `false`?

Eslint should be in all our projects and I'd prefer if we dont let people configure it.

File: packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts [hunk 4]
```diff
@@ -13,6 +15,9 @@ export interface FrontlineJsConfigWebpackPluginOptions {
     mode: "production" | "development" | "none";
     browserslistEnv: string;
     babelConfigFile?: string;
+    tsConfigFile?: string;
+    plugins?: any;
```
Comment (dennishn) [inline packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts:19]: I did not previously add support for adding custom Webpack Plugins through our plugins - mantra was "take ours - it works well and is guarenteed to work, or create your own webpack configuration".

Are there any features this adds?

File: performance/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "compilerOptions": {
+    "resolveJsonModule": true
```
Comment (dennishn) [inline performance/tsconfig.json:3]: This is already configured in the js-config-webpack-plugin package tsconfig, so should not be needed here.

File: packages/js-config-webpack-plugin/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+module.exports = {
```
Comment (dennishn) [inline packages/js-config-webpack-plugin/.eslintrc.js:1]: The individual packages should not have their own specific linting / formatting rules - they should be using the configurations in root.

File: packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts (hunk 4 above)
Comment (dennishn) [inline packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts:19]: Our webpack configuration plugins should be opinionated and not offer any options really. If people wants to create a custom webpack config, they can do so in their webpack.config.js file.

Using our configurations should guarentee a working setup with our opionated, but awesome settings.

File: packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts (hunk 3 above)
Comment (dennishn) [inline packages/js-config-webpack-plugin/src/config/FrontlineJsConfigWebpackPlugin.ts:20]: Linting / Formatting should be handled just by extending on our eslint config.

What is the purpose of also having it as an option for the webpack js/ts transpilation plugin?

File: performance/src/typings.d.ts [hunk 7]
```diff
@@ -0,0 +1,2 @@
+// This will make TypeScript understand CSS Modules
+declare module "*.module.scss";
```
Comment (dennishn) [inline performance/src/typings.d.ts:2]: I think our "default" typings should include a few more module definitions (taken from CRA, adjusted svg file names to match our setup):

```
declare module '*.bmp' {
    const src: string;
    export default src;
}

declare module '*.gif' {
    const src: string;
    export default src;
}

declare module '*.jpg' {
    const src: string;
    export default src;
}

declare module '*.jpeg' {
    const src: string;
    export default src;
}

declare module '*.png' {
    const src: string;
    export default src;
}

declare module '*.webp' {
    const src: string;
    export default src;
}

declare module '*.component.svg' {
    import * as React from 'react';

    export const ReactComponent: React.FunctionComponent<React.SVGProps<
        SVGSVGElement
        > & { title?: string }>;

    const src: string;
    export default src;
}

declare module '*.inline.svg' {
    const src: string;
    export default src;
}

declare module '*.svg' {
    const src: string;
    export default src;
}

declare module '*.module.css' {
    const classes: { readonly [key: string]: string };
    export default classes;
}

declare module '*.module.scss' {
    const classes: { readonly [key: string]: string };
    export default classes;
}
```

**Comment by dennishn** [comment]
I tried renaming the React JS files in `/performance` to `tsx` - and ran into a few issues

- The linter is complaining that the imported React components, and React itself, are never used (even though they are, in the JSX):
```
WARNING in C:\Users\Dennis\Documents\frontline\performance\src\App.tsx(1,8):
@typescript-eslint/no-unused-vars: 'React' is defined but never used.
WARNING in C:\Users\Dennis\Documents\frontline\performance\src\App.tsx(8,8):
@typescript-eslint/no-unused-vars: 'Eager' is defined but never used.
WARNING in C:\Users\Dennis\Documents\frontline\performance\src\App.tsx(10,7):
@typescript-eslint/no-unused-vars: 'Lazy' is assigned a value but never used.
```

- The JSON import had to be updated to include the file extension 

- Even though you have configured tsconfig with ` "allowJs": false` - it still eats the JS files in the performance directory - i dont think it should.


I would like if the project in `performance` includes some sort of Typescript setup, either new additional TS files, or a complete standalone TS example.

**Pull request merged**